src/contracts/core/StakeRootCompendium.sol

```diff
@@ -0,0 +1,618 @@
+        // TODO: verify proof
+        // TODO: prevent race incentives and public mempool sniping, eg embed chargeRecipient in the proof
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function confirmStakeRoot(uint32 index, bytes32 stakeRoot) external {
+        require(msg.sender == rootConfirmer, OnlyRootConfirmerCanConfirm());
+        IStakeRootCompendium.StakeRootSubmission storage stakeRootSubmission = stakeRootSubmissions[index];
+        require(stakeRootSubmission.stakeRoot == stakeRoot, StakeRootDoesNotMatch());
+        require(stakeRootSubmission.confirmed, TimestampAlreadyConfirmed());
+        stakeRootSubmission.confirmed = true;
+    }
+
+    /// SET FUNCTIONS
+
+    /// @inheritdoc IStakeRootCompendium
+    function setMaxChargePerProof(
+        uint96 _maxChargePerProof
+    ) external onlyOwner {
+        require(_maxChargePerProof >= chargePerProof.latest(), MaxTotalChargeMustBeGreaterThanTheCurrentTotalCharge());
+        chargeParams.maxChargePerProof = _maxChargePerProof;
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function setChargePerProof(uint96 _chargePerStrategy, uint96 _chargePerOperatorSet) external onlyOwner {
+        _updateCumulativeCharge();
+        ChargeParams storage params = chargeParams;
+        params.chargePerStrategy = _chargePerStrategy;
+        params.chargePerOperatorSet = _chargePerOperatorSet;
+        _updateChargePerProof();
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function setProofIntervalSeconds(
+        uint32 _proofIntervalSeconds
+    ) external onlyOwner {
+        _updateCumulativeCharge();
+        // we must not interrupt pending proof calculations by rugging the outstanding calculationTimestamps
+        require(
+            stakeRootSubmissions[stakeRootSubmissions.length - 1].calculationTimestamp
+                == cumulativeChargeParams.lastUpdateTimestamp,
+            NoProofsThatHaveBeenChargedButNotSubmitted()
+        );
+        cumulativeChargeParams.proofIntervalSeconds = _proofIntervalSeconds;
+    }
+
+    /// @inheritdoc IStakeRootCompendium
+    function setRootConfirmer(
+        address _rootConfirmer
+    ) public onlyOwner {
+        rootConfirmer = _rootConfirmer;
+    }
+
+    /// INTERNAL FUNCTIONS
+
+    function _removeFromStakeTree(
+        OperatorSet memory operatorSet
+    ) internal {
+        OperatorSet memory substituteOperatorSet = operatorSets[operatorSets.length - 1];
+        Snapshots.History storage snapshotHistory = operatorSetToIndex[operatorSet.avs][operatorSet.operatorSetId];
+        uint224 operatorSetIndex = snapshotHistory.latest();
+        operatorSets[operatorSetIndex] = substituteOperatorSet;
+        operatorSets.pop();
+
+        // update the index of the operator sets
+        // note when there is only one operator set left, the index will not be updated as the operator set will be removed in the next step
```

Comment:
Or even if the `operatorSet` passed in to call data is in fact the last operatorSet in the `opeartorSets` array (ie. equal to `substituteOperatorSet`. There doesn't necessarily have to be one operator set left. 

Why can't we just simplify this logic to if `operatorSet == substituteOperatorSet` then you only need to push `REMOVED_INDEX`and pop... else do previous in addition to handling substitute

Reply:
didn't even think about this case. agree it should be handled gracefully